src/main/java/org/logstash/input/SinceDB.java

```diff
@@ -0,0 +1,118 @@
+            return false;
+        }
+
+        @Override
+        public void flush() {
+            // unassigned SinceDB hasn't anything to flush
+        }
+
+        @Override
+        public Path getCurrentSegment() {
+            throw new IllegalStateException("Unassigned SinceDB doesn't have currentSegment");
+        }
+
+        @Override
+        public long getOffset() {
+            throw new IllegalStateException("Unassigned SinceDB doesn't have offset");
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(SinceDB.class);
+
+    private final Path sinceDbPath;
+    private final Path currentSegment;
+    private final long offset;
+
+    private SinceDB(Path sinceDbPath, Path currentSegment, long offset) {
+        this.sinceDbPath = sinceDbPath;
+        this.currentSegment = currentSegment;
+        this.offset = offset;
+    }
+
+    static SinceDB load(Path sinceDbPath) throws IOException {
+        byte[] bytes = Files.readAllBytes(sinceDbPath);
+        if (bytes.length == 0) {
+            return createUnassigned(sinceDbPath);
+        }
+        ByteBuffer buffer = ByteBuffer.wrap(bytes);
+        verifyVersion(buffer);
+        Path segmentPath = decodePath(buffer);
+        long offset = buffer.getLong();
+
+        return new SinceDB(sinceDbPath, segmentPath, offset);
+    }
+
+    private static Path decodePath(ByteBuffer buffer) {
+        int segmentPathStringLength = buffer.getInt();
+        byte[] segmentPathBytes = new byte[segmentPathStringLength];
+        buffer.get(segmentPathBytes);
+        return Paths.get(new String(segmentPathBytes));
+    }
+
+    private static void verifyVersion(ByteBuffer buffer) {
+        char version = buffer.getChar();
+        if (DeadLetterQueueInputPlugin.VERSION != version) {
+            throw new RuntimeException("Sincedb version:" + version + " does not match: " + DeadLetterQueueInputPlugin.VERSION);
+        }
+    }
+
+    static SinceDB createUnassigned(Path sinceDbPath) {
+        return new UnassignedSinceDB(sinceDbPath);
+    }
+
+    public void flush() {
+        if (currentSegment == null) {
+            return;
+        }
+        logger.debug("Flushing DLQ last read position");
+        String path = currentSegment.toAbsolutePath().toString();
+        ByteBuffer buffer = ByteBuffer.allocate(path.length() + 1 + 64);
+        buffer.putChar(DeadLetterQueueInputPlugin.VERSION);
+        buffer.putInt(path.length());
+        buffer.put(path.getBytes());
+        buffer.putLong(offset);
+        try {
+            Files.write(sinceDbPath, buffer.array());
+        } catch (IOException e) {
+            logger.error("failed to write DLQ offset state to " + sinceDbPath, e);
+        }
+    }
+
+    public Path getCurrentSegment() {
+        return currentSegment;
+    }
+
+    public long getOffset() {
+        return offset;
+    }
+
+    public SinceDB updatePosition(DeadLetterQueueReader reader) {
```

Comment:
I'm not sure about this approach - it threw me for a loop when I was trying to figure out how an "assigned" `SinceDB` object came to be created.

At first glance, this looks like a mutation method on the `SinceDB` object, rather than a method that constructs a new object, and _requires_ the caller to use this newly created version of the object. 
Calling `updatePosition` without using the return value would not get caught and could introduce some subtle bugs when expanding the use of `updatePosition` outside of the `close` method.

Reply:
You a right, it was a mutation method, then I switched SinceDB to be a value object without changing it. I'll provide a better name.

Reply:
Renamed to `retrieveNewPosition`.

Reply:
Is there a reason why you moved from the Optional usage to the Null object pattern? The Null Object pattern here introduces some complexity, including the use of a tricky inheritance, which requires constructing the superclass with "bogus" parameters, such as `Paths.get(System.getProperty("user.home")`, and requires this method that "looks" like a mutation?

Reply:
The motivation was the using the `Optional<SinceDB>` as a field, forced me to check if the value is present to execute some actions or not. For example, with optional the `sinceDb.flush()` has to be surrounded with presence check:
```java
finally {
  if (sinceDb.isPresent() {
    sinceDb.get().flush();
  }
}
```
becomes more straightforward with the Null parttern
```java
finally {
  sinceDb.flush();
}
```

Reading through the `Optional` [javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Optional.html) it's intended to be used as method return type where there is the concept of "no result".
Having a field of a class `Optional<>` , when we can't use the `ifPresent` and `map` transformation methods, for to have a lot of `optional_instance.get()` calls.

So with the Null pattern we moved the if-present logic from `if`s to method dispatch, and think provide more context knowledge then the Optional.

Reply:
I also thought that having the `SinceDB` as non mutating class could help us when in the plugin we have concurrent update of it, and this should happen when we save the sinceDB from a scheduler thread.

Reply:
I wonder if it would make sense to have a `SinceDB` interface with static methods to retrieve a `loaded` sinceDB and to "update" an existing sinceDB, with concrete implementations of AssignedDB and UnassignedDB. This would remove the need to populate a constructor with "fake" values, and would remove the need to have an instance method on SinceDB, which requires the user of the class to know that the class is intended to replace the existing instance:

Something along the lines of 

```
interface SinceDB {

  Path getPath();
  void flush();
  Path getCurrentSegment();
  long getOffset();
  boolean isAssigned();

  /**
   * Load SinceDB from given path
   * @param Path of sinceDB
   * @return new instance of SinceDB. This instance may be unassigned if the path does not exist, or has no entries written 
   *
   */
  static SinceDB fromPath(Path sinceDbPath) throws IOException;

  /**
   * Create new SinceDB instance from the position of the reader.
   *
   * @param instance of SinceDB to update
   * @param reader the reader that needs to be queried.
   * @return new instance of SinceDB containing the new position.
   */
  static SinceDB getUpdated(SinceDB oldSinceDb, DeadLetterQueueReader reader);
}
```

The lack of private static methods in interfaces in Java 8 makes this more annoying than it would be in Java 11 though...